### PR TITLE
Dcerpc incomplete 5699/v5

### PR DESCRIPTION
**Pull request by inashivb**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket: https://redmine.openinfosecfoundation.org/issues/5699

Previous PR: #9276

Changes since v4:
- review comments addressed
- rebased on latest master

### CODE REVIEW

**Comment by suricata-qa** [comment]
WARNING:
 
 | field | baseline | test | % |
|--|--|--|--| 
| | SURI_TLPR1_stats_chk | | |
| .app_layer.flow.dcerpc_tcp | 4185 | 40 | 0.96% |
| .app_layer.tx.dcerpc_tcp | 5949 | 4165 | 70.01% |
| .app_layer.error.dcerpc_tcp.parser | 6248 | 10393 | 166.34% | 

Pipeline 22251

**Comment by inashivb** [comment]
> WARNING:
> field 	baseline 	test 	%
> 	SURI_TLPR1_stats_chk 		
> .app_layer.flow.dcerpc_tcp 	4185 	40 	0.96%
> .app_layer.tx.dcerpc_tcp 	5949 	4165 	70.01%
> .app_layer.error.dcerpc_tcp.parser 	6248 	10393 	166.34%
> 
> Pipeline 22251

Based on the smaller pcap of this given by Corey, my analysis is that these stat deviations are likely correct. This is because of the following.
|`master`|`this PR`|
|-------------|------------|
|Invalid header -> returns `Ok` from the parser|Invalid header -> returns `Error` from the parser|

and there's a lot of traffic which is not `dcerpc` and has invalid header.
Most parsing errors say `Major version: 77 Minor version: 97` and return Ok.
For the reference, valid header should have: `Major version: 5 Minor version: 0` for `dcerpc/tcp`..

File: rust/src/dcerpc/dcerpc.rs [hunk 1]
```diff
@@ -1811,216 +1811,6 @@ mod tests {
...
-            0x62, 0x9f, 0x3d, 0x8e, 0x47, 0x73, 0x02, 0x00, 0x02, 0x00, 0x04, 0x5d, 0x88, 0x8a,
-            0xeb, 0x1c, 0xc9, 0x11, 0x9f, 0xe8, 0x08, 0x00, 0x2b, 0x10, 0x48, 0x60, 0x02, 0x00,
-            0x00, 0x00, 0x0f, 0x00, 0x01, 0x00, 0xc8, 0x4f, 0x32, 0x4b, 0x70, 0x16, 0xd3, 0x01,
-            0x12, 0x78, 0x5a, 0x47, 0xbf, 0x6e, 0xe1, 0x88, 0x03, 0x00, 0x00, 0x00, 0x04, 0x5d,
-            0x88, 0x8a, 0xeb, 0x1c, 0xc9, 0x11, 0x9f, 0xe8, 0x08, 0x00, 0x2b, 0x10, 0x48, 0x60,
-            0x02, 0x00, 0x00, 0x00,
-        ];
-        let mut dcerpc_state = DCERPCState::new();
-        assert_eq!(
-            AppLayerResult::ok(),
```
Comment (catenacyber) [inline rust/src/dcerpc/dcerpc.rs:1910]: Can we not change the expectation that it is not ok but incomplete ? 

And append the next buffer to it ?

Comment (inashivb) [reply rust/src/dcerpc/dcerpc.rs:1910]: Can do that indeed. The code already has a good coverage though w the s-v tests and existing unit tests, do you think it adds value? Please lmk. Thank you!

Comment (catenacyber) [reply rust/src/dcerpc/dcerpc.rs:1910]: Ok, but then you need to rewrite the commit message

> With the introduction of AppLayerResult::incomplete API, fragmented data
is no longer handled fully in the dcerpc code making the tests about
such data invalid.

The tests are still valid, but they are now done with suricata-verify test abc or xyz

Comment (inashivb) [reply rust/src/dcerpc/dcerpc.rs:1910]: Will do. Thank you!

**Comment by catenacyber** [comment]
> > WARNING:
> > field 	baseline 	test 	%
> > SURI_TLPR1_stats_chk
> > .app_layer.flow.dcerpc_tcp 	4185 	40 	0.96%
> > .app_layer.tx.dcerpc_tcp 	5949 	4165 	70.01%
> > .app_layer.error.dcerpc_tcp.parser 	6248 	10393 	166.34%
> > Pipeline 22251
> 
> Based on the smaller pcap of this given by Corey, my analysis is that these stat deviations are likely correct. This is because of the following.
> 
> `master`	`this PR`
> Invalid header -> returns `Ok` from the parser	Invalid header -> returns `Error` from the parser
> and there's a lot of traffic which is not `dcerpc` and has invalid header. Most parsing errors say `Major version: 77 Minor version: 97` and return Ok. For the reference, valid header should have: `Major version: 5 Minor version: 0` for `dcerpc/tcp`..
> 
> Edit: Something is off. If there was a version mismatch in the header, the probing itself should have failed.. **Update 1**: Probing does not seem to run for every input. **Update 2**: The pcap contains SMB/DCERPC traffic so not sure why this parser is even invoked. The parser however seems to work correctly. It receives in order:
> 
> * A valid `DCERPC` pkt
> * Trailing data from `NBSS` (This is rejected by the parser as the header validation fails. This is correct.)

Do you have SV test for this ?

**Comment by catenacyber** [comment]
Setting the label `[needs baseline update](https://github.com/OISF/suricata/labels/needs%20baseline%20update)`

**Review by catenacyber — CHANGES_REQUESTED** [review]
This looks really good.

Should the commit [dcerpc: return error on invalid header](https://github.com/OISF/suricata/pull/11659/commits/f7a7fd2fa554f0d54f58ded04e7d6b556ba3e406) be its own first preliminary PR and ticket ? As it brings baseline changes

**Comment by inashivb** [comment]
> This looks really good.
> 
> Should the commit [dcerpc: return error on invalid header](https://github.com/OISF/suricata/pull/11659/commits/f7a7fd2fa554f0d54f58ded04e7d6b556ba3e406) be its own first preliminary PR and ticket ? As it brings baseline changes

Done in https://github.com/OISF/suricata/pull/11676

**Pull request closed without merging**